fluss-server/src/main/java/org/apache/fluss/server/DynamicServerConfig.java

```diff
@@ -115,68 +148,296 @@ boolean isAllowedConfig(String key) {
+        effectiveChanges.forEach(
                 (key, value) -> {
                     if (value == null) {
-                        props.remove(key);
+                        configMap.remove(key);
                     } else {
-                        props.put(key, value);
+                        configMap.put(key, value);
                     }
                 });
+        return configMap;
+    }
+
+    /** Updates internal state after successful reconfiguration. */
+    private void updateInternalState(
+            Configuration newConfig,
+            Map<String, String> newConfigMap,
+            Map<String, String> newDynamicConfigs) {
+        currentConfig = newConfig;
+        currentConfigMap.clear();
+        currentConfigMap.putAll(newConfigMap);
+        dynamicConfigs.clear();
+        dynamicConfigs.putAll(newDynamicConfigs);
+    }
+
+    /**
+     * Validates a single config entry including type parsing and business validation.
+     *
+     * @param configKey config key
+     * @param oldValueStr old value string
+     * @param newValueStr new value string
+     * @throws ConfigException if validation fails
+     */
+    private void validateSingleConfig(String configKey, String oldValueStr, String newValueStr)
+            throws ConfigException {
+        // Get ConfigOption for type information
+        ConfigOption<?> configOption = ConfigOptions.getConfigOption(configKey);
+
+        // For configs with allowed prefixes (like "datalake."), skip ConfigOption validation
+        // and rely on ServerReconfigurable's business validation
+        boolean hasPrefixConfig = false;
+        for (String prefix : ALLOWED_CONFIG_PREFIXES) {
+            if (configKey.startsWith(prefix)) {
+                hasPrefixConfig = true;
+                break;
+            }
+        }
+
+        if (configOption == null && !hasPrefixConfig) {
+            throw new ConfigException(
+                    String.format("No ConfigOption found for config key: %s", configKey));
+        }
+
+        // Parse and validate type only if ConfigOption exists
+        Object newValue = null;
+        if (configOption != null && newValueStr != null) {
+            Configuration tempConfig = new Configuration();
+            tempConfig.setString(configKey, newValueStr);
+            try {
+                newValue = tempConfig.getOptional(configOption).get();
+            } catch (Exception e) {
+                throw new ConfigException(
+                        String.format(
+                                "Cannot parse '%s' as %s for config '%s'",
+                                newValueStr,
+                                configOption.isList()
+                                        ? "List<" + configOption.getClazz().getSimpleName() + ">"
+                                        : configOption.getClazz().getSimpleName(),
+                                configKey),
+                        e);
```

Comment:
Add the exception message into `ConfigException` top error message, because `ApiException` will truncate the exception stack and will hidden the root cause.